File: vane-regions/src/main/java/org/oddlama/vane/regions/event/RegionPortalRoleSettingEnforcer.java
package org.oddlama.vane.regions.event;

import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.oddlama.vane.core.Listener;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.portals.event.PortalActivateEvent;
import org.oddlama.vane.portals.event.PortalChangeSettingsEvent;
import org.oddlama.vane.portals.event.PortalConstructEvent;
import org.oddlama.vane.portals.event.PortalDeactivateEvent;
import org.oddlama.vane.portals.event.PortalDestroyEvent;
import org.oddlama.vane.portals.event.PortalLinkConsoleEvent;
import org.oddlama.vane.portals.event.PortalOpenConsoleEvent;
import org.oddlama.vane.portals.event.PortalSelectTargetEvent;
import org.oddlama.vane.portals.event.PortalUnlinkConsoleEvent;
import org.oddlama.vane.regions.Regions;
import org.oddlama.vane.regions.region.RoleSetting;

public class RegionPortalRoleSettingEnforcer extends Listener<Regions> {

    public RegionPortalRoleSettingEnforcer(Context<Regions> context) {
        super(context);
    }

    public boolean checkSettingAt(
        final Location location,
        final Player player,
        final RoleSetting setting,
        final boolean checkAgainst
    ) {
        final var region = getModule().regionAt(location);
        if (region == null) {
            return false;
        }

        final var group = region.regionGroup(getModule());
        return group.getRole(player.getUniqueId()).getSetting(setting) == checkAgainst;
    }

    public boolean checkSettingAt(
        final Block block,
        final Player player,
        final RoleSetting setting,
        final boolean checkAgainst
    ) {
        final var region = getModule().regionAt(block);
        if (region == null) {
            return false;
        }

        final var group = region.regionGroup(getModule());
        return group.getRole(player.getUniqueId()).getSetting(setting) == checkAgainst;
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPortalActivate(final PortalActivateEvent event) {
        if (event.getPlayer() == null) {
            // Activated by redstone -> Always allow. It's the job of the region
            // owner to prevent redstone interactions if a portal shouldn't be activated.
            return;
        }

        if (checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.PORTAL, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPortalDeactivate(final PortalDeactivateEvent event) {
        if (checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.PORTAL, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPortalConstruct(final PortalConstructEvent event) {
        // We have to check all blocks here, because otherwise players
        // could "steal" boundary blocks from unowned regions
        for (final var block : event.getBoundary().allBlocks()) {
            // Portals in regions may only be constructed by region administrators
            if (checkSettingAt(block, event.getPlayer(), RoleSetting.ADMIN, false)) {
                event.setCancelled(true);
                return;
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPortalDestroy(final PortalDestroyEvent event) {
        if (event.getPortal().owner().equals(event.getPlayer().getUniqueId())) {
            // Owner may always use their portals
            return;
        }

        // We do NOT have to check all blocks here, because
        // an existing portal with its spawn inside a region
        // that the player controls can be considered proof of authority.
        if (checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.ADMIN, false)) {
            // Portals in regions may only be destroyed by region administrators
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = false)
    public void onPortalLinkConsole(final PortalLinkConsoleEvent event) {
        if (event.getPortal() != null && event.getPortal().owner().equals(event.getPlayer().getUniqueId())) {
            // Owner may always use their portals
            return;
        }

        if (event.getPortal() != null && getModule().regionAt(event.getPortal().spawn()) != null) {
            // Portals in regions may be administrated by region administrators,
            // not only be the owner
            event.setCancelIfNotOwner(false);
        }

        // Portals in regions may only be administrated by region administrators
        // Check permission on console
        if (checkSettingAt(event.getConsole(), event.getPlayer(), RoleSetting.ADMIN, false)) {
            event.setCancelled(true);
            return;
        }

        // Check permission on portal if any
        if (
            event.getPortal() != null &&
            checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.ADMIN, false)
        ) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = false)
    public void onPortalUnlinkConsole(final PortalUnlinkConsoleEvent event) {
        if (event.getPortal().owner().equals(event.getPlayer().getUniqueId())) {
            // Owner may always use their portals
            return;
        }

        if (getModule().regionAt(event.getPortal().spawn()) != null) {
            // Portals in regions may be administrated by region administrators,
            // not only be the owner
            event.setCancelIfNotOwner(false);
        }

        // Portals in regions may only be administrated by region administrators
        // Check permission on console
        if (checkSettingAt(event.getConsole(), event.getPlayer(), RoleSetting.ADMIN, false)) {
            event.setCancelled(true);
            return;
        }

        // Check permission on portal
        if (checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.ADMIN, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPortalOpenConsole(final PortalOpenConsoleEvent event) {
        if (event.getPortal().owner().equals(event.getPlayer().getUniqueId())) {
            // Owner may always use their portals
            return;
        }

        // Check permission on console
        if (checkSettingAt(event.getConsole(), event.getPlayer(), RoleSetting.PORTAL, false)) {
            event.setCancelled(true);
            return;
        }

        // Check permission on portal
        if (checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.PORTAL, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPortalSelectTarget(final PortalSelectTargetEvent event) {
        if (event.getPortal().owner().equals(event.getPlayer().getUniqueId())) {
            // Owner may always use their portals
            return;
        }

        // Check permission on source portal
        if (checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.PORTAL, false)) {
            event.setCancelled(true);
            return;
        }

        // Check permission on target portal
        if (
            event.getTarget() != null &&
            checkSettingAt(event.getTarget().spawn(), event.getPlayer(), RoleSetting.PORTAL, false)
        ) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = false)
    public void onPortalChangeSettings(final PortalChangeSettingsEvent event) {
        if (event.getPortal().owner().equals(event.getPlayer().getUniqueId())) {
            // Owner may always use their portals
            return;
        }

        if (getModule().regionAt(event.getPortal().spawn()) == null) {
            return;
        }

        // Portals in regions may be administrated by region administrators,
        // not only be the owner
        event.setCancelIfNotOwner(false);

        // Now check if the player has the permission
        if (checkSettingAt(event.getPortal().spawn(), event.getPlayer(), RoleSetting.ADMIN, false)) {
            event.setCancelled(true);
        }
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/event/RegionPortalRoleSettingEnforcer.kt
package org.oddlama.vane.regions.event

import org.bukkit.Location
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.oddlama.vane.core.Listener
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.portals.event.*
import org.oddlama.vane.regions.Regions
import org.oddlama.vane.regions.region.RoleSetting

/**
 * Enforces region role permissions for vane-portals interactions.
 */
class RegionPortalRoleSettingEnforcer(context: Context<Regions?>?) : Listener<Regions?>(context) {
    /**
     * Owning regions module instance.
     */
    private val regions: Regions
        get() = requireNotNull(module)

    /**
     * Checks whether a role setting at a location matches the expected value.
     */
    fun checkSettingAt(
        location: Location,
        player: Player,
        setting: RoleSetting,
        checkAgainst: Boolean
    ): Boolean {
        /** Region at the queried location. */
        val region = regions.regionAt(location) ?: return false

        /** Region group owning that region. */
        val group = region.regionGroup(regions) ?: return false

        /** Effective player role inside the region group. */
        val role = group.getRole(player.uniqueId) ?: return false
        return role.getSetting(setting) == checkAgainst
    }

    /**
     * Checks whether a role setting at a block matches the expected value.
     */
    fun checkSettingAt(
        block: Block,
        player: Player,
        setting: RoleSetting,
        checkAgainst: Boolean
    ): Boolean {
        /** Region containing the queried block. */
        val region = regions.regionAt(block) ?: return false

        /** Region group owning that region. */
        val group = region.regionGroup(regions) ?: return false

        /** Effective player role inside the region group. */
        val role = group.getRole(player.uniqueId) ?: return false
        return role.getSetting(setting) == checkAgainst
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
            /**
             * Cancels portal activation when `PORTAL` usage is denied.
             */
    fun onPortalActivate(event: PortalActivateEvent) {
        val player = event.player ?: run {
            // Activated by redstone -> Always allow. It's the job of the region
            // owner to prevent redstone interactions if a portal shouldn't be activated.
            return
        }
        val portal = event.portal ?: return

        if (checkSettingAt(portal.spawn(), player, RoleSetting.PORTAL, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
            /**
             * Cancels portal deactivation when `PORTAL` usage is denied.
             */
    fun onPortalDeactivate(event: PortalDeactivateEvent) {
        val portal = event.portal ?: return
        val player = event.player ?: return
        if (checkSettingAt(portal.spawn(), player, RoleSetting.PORTAL, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
            /**
             * Restricts portal construction to region admins for all boundary blocks.
             */
    fun onPortalConstruct(event: PortalConstructEvent) {
        val boundary = event.boundary ?: return
        val player = event.player ?: return
        // We have to check all blocks here, because otherwise players
        // could "steal" boundary blocks from unowned regions
        for (block in boundary.allBlocks()) {
            // Portals in regions may only be constructed by region administrators
            if (checkSettingAt(block, player, RoleSetting.ADMIN, false)) {
                event.isCancelled = true
                return
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
            /**
             * Restricts portal destruction to owners or region admins.
             */
    fun onPortalDestroy(event: PortalDestroyEvent) {
        if (event.portal.owner() == event.player.uniqueId) {
            // Owner may always use their portals
            return
        }

        // We do NOT have to check all blocks here, because
        // an existing portal with its spawn inside a region
        // that the player controls can be considered proof of authority.
        if (checkSettingAt(event.portal.spawn(), event.player, RoleSetting.ADMIN, false)) {
            // Portals in regions may only be destroyed by region administrators
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = false)
            /**
             * Restricts linking consoles to portals based on region admin permissions.
             */
    fun onPortalLinkConsole(event: PortalLinkConsoleEvent) {
        val player = event.player
        val portal = event.portal
        if (portal != null && portal.owner() == player.uniqueId) {
            // Owner may always use their portals
            return
        }

        if (portal != null && regions.regionAt(portal.spawn()) != null) {
            // Portals in regions may be administrated by region administrators,
            // not only be the owner
            event.setCancelIfNotOwner(false)
        }

        // Portals in regions may only be administrated by region administrators
        // Check permission on console
        val console = event.console ?: return
        if (checkSettingAt(console, player, RoleSetting.ADMIN, false)) {
            event.isCancelled = true
            return
        }

        // Check permission on portal if any
        if (portal != null && checkSettingAt(portal.spawn(), player, RoleSetting.ADMIN, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = false)
            /**
             * Restricts unlinking consoles from portals based on region admin permissions.
             */
    fun onPortalUnlinkConsole(event: PortalUnlinkConsoleEvent) {
        val player = event.player
        val portal = event.portal
        if (event.portal.owner() == event.player.uniqueId) {
            // Owner may always use their portals
            return
        }

        if (regions.regionAt(portal.spawn()) != null) {
            // Portals in regions may be administrated by region administrators,
            // not only be the owner
            event.setCancelIfNotOwner(false)
        }

        // Portals in regions may only be administrated by region administrators
        // Check permission on console
        val console = event.console ?: return
        if (checkSettingAt(console, player, RoleSetting.ADMIN, false)) {
            event.isCancelled = true
            return
        }

        // Check permission on portal
        if (checkSettingAt(portal.spawn(), player, RoleSetting.ADMIN, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
            /**
             * Restricts opening portal consoles based on `PORTAL` permissions.
             */
    fun onPortalOpenConsole(event: PortalOpenConsoleEvent) {
        val portal = event.portal ?: return
        val player = event.player ?: return
        if (portal.owner() == player.uniqueId) {
            // Owner may always use their portals
            return
        }

        // Check permission on console
        val console = event.console ?: return
        if (checkSettingAt(console, player, RoleSetting.PORTAL, false)) {
            event.isCancelled = true
            return
        }

        // Check permission on portal
        if (checkSettingAt(portal.spawn(), player, RoleSetting.PORTAL, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
            /**
             * Restricts selecting portal targets based on `PORTAL` permissions.
             */
    fun onPortalSelectTarget(event: PortalSelectTargetEvent) {
        val portal = event.portal ?: return
        val player = event.player ?: return
        if (portal.owner() == player.uniqueId) {
            // Owner may always use their portals
            return
        }

        // Check permission on source portal
        if (checkSettingAt(portal.spawn(), player, RoleSetting.PORTAL, false)) {
            event.isCancelled = true
            return
        }

        // Check permission on target portal
        val target = event.target
        if (target != null && checkSettingAt(target.spawn(), player, RoleSetting.PORTAL, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = false)
            /**
             * Restricts portal settings changes to owners or region admins.
             */
    fun onPortalChangeSettings(event: PortalChangeSettingsEvent) {
        if (event.portal.owner() == event.player.uniqueId) {
            // Owner may always use their portals
            return
        }

        if (regions.regionAt(event.portal.spawn()) == null) {
            return
        }

        // Portals in regions may be administrated by region administrators,
        // not only be the owner
        event.setCancelIfNotOwner(false)

        // Now check if the player has the permission
        if (checkSettingAt(event.portal.spawn(), event.player, RoleSetting.ADMIN, false)) {
            event.isCancelled = true
        }
    }
}
